### ISSUE TITLE
The `msize` opcode in SeaEVM returns an incorrect value

### ISSUE THREAD

**Issue by Alleysira** [SealSC/SealEVM#28]
# The `msize` opcode in SeaEVM returns an incorrect value

Hello @AKACoder, I found another opcode bug in the SealEVM repository at https://github.com/SealSC/SealEVM/blob/cadd48e72a519afe7ce73ee2cb07be07d33c9ee2/memory/memory.go#L129 This bug causes SealEVM to output wrong memory size with other standard EVM implementations, which is against the Ethereum virtual machine opcode definition.

# Bug description

When executing the opcode 0x59 `msize`, SealEVM will directly return the actual memory length, like `int64(len(m.cell))`, but according to [evm.code](https://www.evm.codes/?fork=cancun) and other EVMs' implementation, the return value in stack should be a multiple of a word (32 bytes). 

# How to reproduce

## Set deployed bytecode

Simply pull the latest version of the SealEVM and enter the `example` dir.

```shell
git clone https://github.com/SealSC/SealEVM.git
cd example
nano code.go
```

Change the ` var deployCode` in `example/code.go` to bytecode `631456015261af0b525960005260406000f3`

```
 var deployCode, _ = hex.DecodeString("631456015261af0b525960005260406000f3")
```

Where `631456015261af0b525960005260406000f3` are the opcodes as shown below:

```
PUSH4 0x14560152
PUSH2 0xaf0b
MSTORE 
MSIZE
PUSH1 0x40
PUSH1 0x00
RETURN
```
## Add logging statement

Then add a line of  log statement  in `example/example.go ` to print the result data.

```go
import "log"

	ret, err := evm.ExecuteContract(false)
	log.Print(hex.EncodeToString(ret.ResultData))
	//check error
```

## Run

```
cd example
go build -o sealevm
./sealevm
```

You can see that the msize returnd 0xaf2b.

## Result

```shell
SealEVM/example$ ./sealevm
2024/07/27 19:15:24 000000000000000000000000000000000000000000000000000000000000af2b0000000000000000000000000000000000000000000000000000000000000000
counter:
counter:
```

Here is the result [gethout.json](https://github.com/user-attachments/files/16400089/gethout.json) from `geth/evm 1.14.5-stable-0dd173a7`. Geth and other EVM returned `0xaf40`. The result of SealEVM is [sealout.json](https://github.com/user-attachments/files/16400099/sealout.json).

```
// geth
000000000000000000000000000000000000000000000000000000000000af400000000000000000000000000000000000000000000000000000000000000000
// SealEVM 
000000000000000000000000000000000000000000000000000000000000af2b0000000000000000000000000000000000000000000000000000000000000000
```

However, SealEVM returned `0xaf2b` , not a multiple of 32.

```
>>> 0xaf40/32
1402.0
>>> 0xaf2b/32
1401.34375
```

**Comment by AKACoder** [comment]
@Alleysira Thank you for your thoughtful review of SealEVM! 

However, regarding what you mentioned in this issue, the return value of MSIZE is always a multiple of a word (32 bytes). This is because EVM’s memory expands in units of 32 bytes. 
The yellow paper also explains the organization of EVM’s memory as: "**This memory is organized into 32-byte words.**" 

Therefore, I plan to modify the related methods of the Memory structure in the memory.go file to make the memory organization and expansion of SealEVM comply with the description in the yellow paper, and also pave the way for trying to keep up with the standard gas consumption of EVM in the future.

Once again, thank you for the attention to SealEVM!

**Comment by Alleysira** [comment]
Thanks for your quick response!

The memory issue is precisely the concern I wish to point out.  And you mentioned 

>  keep up with the standard gas consumption of EVM

I think the Ethereum execution specs at [gas.py](https://github.com/ethereum/execution-specs/blob/44ceab258abd42fdebdee06996bf0700b205ef98/src/ethereum/cancun/vm/gas.py#L154-L190) can be of help. 

Thanks.

**Comment by AKACoder** [comment]
This issue has been resolved in this commit [4f62b25](https://github.com/SealSC/SealEVM/commit/4f62b25e220bd508374c87f2b667bb81dd91457c). 

> I think the Ethereum execution specs at [gas.py](https://github.com/ethereum/execution-specs/blob/44ceab258abd42fdebdee06996bf0700b205ef98/src/ethereum/cancun/vm/gas.py#L154-L190) can be of help.

And thank you for providing the documentation. I will start a new branch to develop features related to GAS.

@Alleysira
